### PR TITLE
Visual Diffing Tests and Github Action

### PR DESCRIPTION
**Pull request by NicholasHallman**
Following the instructions from https://github.com/BrightspaceUI/visual-diff, https://github.com/BrightspaceUI/actions/tree/master/visual-diff, and https://github.com/Brightspace/iam-build-tokens/blob/master/docs/howto-githubactions.md This PR adds the Visual Diff testing GitHub Action to the engagement dashboard along with 8 different tests

Visual Diff Tests
 - Engagement Dashboard
   - Mobile
   - Desktop
 - User Drill
   - Mobile
   - Desktop    
 - Settings
    - Table Settings
      - Mobile
      - Desktop
    - Chart Settings
      - Mobile
      - Desktop

Recommendations for future tests
 - Summary Card Layout (1, 2, 3, and 4 cards)

FYI: @anhill-D2L @hyehayes @MykolaGalian @Vitalii-Misechko @rohitvinnakota

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -117,7 +119,18 @@ npm run test:diff -- -g some-pattern
 npm run test:diff:golden
 ```
 
-Golden snapshots in source control must be updated by Travis CI. To trigger an update, press the "Regenerate Goldens" button in the pull request `visual-difference` test run.
+A useful way to debug visual tests is to check the report that is created when the tests fail. The reports are an html file. Locally, VSCode users can use an html preview extension for quick debugging. The difference between the goldens and the most recent test are highlighted in red.
+
+Golden snapshots generated locally will not replace the ones generated by the CI Action, to replace goldens please follow the steps bellow.
+
+The github bot may create a new pull request to merge new goldens into your PR. You can check the difference between the goldens and the new snapshots by
```
Comment (johngwilkinson) [inline README.md:126]: Thanks for the docs! I'm not sure what "may create" means here. Does the bot create a PR whenever a vdiff test fails? If not, how do I get it to create the PR?

File: package.json [hunk 2]
```diff
@@ -22,7 +22,7 @@
   "license": "Apache-2.0",
   "devDependencies": {
     "@brightspace-ui/stylelint-config": "0.0.3",
-    "@brightspace-ui/visual-diff": "^1.4.1",
+    "@brightspace-ui/visual-diff": "^2",
```
Comment (bemailloux) [inline package.json:25]: We don't need this dependency or puppeteer anymore - see https://github.com/BrightspaceUI/visual-diff/blob/master/README.md
> This package is designed to be used alongside the visual-diff github action. That action will handle installing so you don't need to include visual-diff and puppeteer in your repo's devDependencies.

EDIT: but I guess you wouldn't get syntax highlighting when writing visual diff tests

File: test/visual-diff/engagement-dashboard.visual-diff.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+			width: 1275,
+			height: 2200,
+			deviceScaleFactor: 1
+		});
+		await page.goto(
+			`${visualDiff.getBaseUrl()}/test/visual-diff/d2l-insights-engagement-dashboard.visual-diff.html`,
+			{ waitUntil: ['networkidle0', 'load'] }
+		);
+
+		await page.keyboard.press('Tab');
```
Comment (johngwilkinson) [inline test/visual-diff/engagement-dashboard.visual-diff.js:23]: what does tab-enter do? code comment?

File: test/visual-diff/settings.visual-diff.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	before(async() => {
+		browser = await puppeteer.launch();
+		page = await visualDiff.createPage(browser);
+		await page.setViewport({
+			width: 1275,
+			height: 2300,
+			deviceScaleFactor: 1
+		});
+		await page.goto(
+			`${visualDiff.getBaseUrl()}/test/visual-diff/d2l-insights-engagement-dashboard.visual-diff.html?v=settings%2C0&ouf=3`,
```
Comment (johngwilkinson) [inline test/visual-diff/settings.visual-diff.js:19]: I did not forsee this use case for the url state, but I'm very happy about it. :)

**Review by johngwilkinson — COMMENTED** [review]
I'm slightly unclear on the `.png`s: are they supposed to be in this PR? The readme says local goldens won't replace the CI ones - is that because CI stores them somewhere else? In which case, why do we also include them in a PR like this? Should they be in `.gitignore`?

File: package.json (hunk 2 above)
Comment (NicholasHallman) [reply package.json:25]: Yeah the linter will get upset if it can't find the deps

**Review by johngwilkinson — APPROVED** [review]
Thanks, very clear now!

**Comment by bemailloux** [comment]
Can you add a few instructions on how to run visual diffs locally? I just did it and there's a tiny bit of setup required - here's what I had to do
```
export GITHUB_REPOSITORY=Brightspace/insights-engagement-dashboard
npm run test:diff:golden
# make some changes
npm run test:diff
```

Just mentioning it because it was _glorious_ to finally see `npm test` actually pass for once.

**Review by bemailloux — APPROVED** [review]
👍

**Pull request merged**